composer/datasets/streaming/dataset.py

```diff
@@ -0,0 +1,385 @@
+            part_max_id (int): Maximum sample ID of this partition.
+        """
+        shards = list(shards)
+        if self.shuffle:
+            np.random.shuffle(shards)
+        for shard in shards:
+            basename = get_shard_basename(shard)
+            self._download_if_missing(basename)
+            shards = shard,
+            self._load_shards(shards, part_min_id, part_max_id)
+        self._done_loading()
+
+    def __len__(self) -> int:
+        """Get the length of the dataset.
+
+        Returns:
+            int: Dataset length.
+        """
+        return self.index.total_samples
+
+    def __getitem__(self, idx: int) -> Any:
+        """Get the sample at the index, assuming its shard is loaded.
+
+        Do not call this directly unless all shards have been loaded. Will crash
+        if the shard is not loaded.
+
+        Args:
+            idx (int): Sample ID.
+
+        Returns:
+            Any: The sample.
+        """
+        shard = self.index.sample_shards[idx]
+        offset = self.index.sample_shard_offsets[idx]
+        size = self.index.bytes_per_sample[idx]
+        fp = self._files[shard]
+        assert fp is not None, 'Tried to __getitem__ a sample that was not loaded.'
+        fp.seek(offset)
+        data = fp.read(size)
+        return bytes_to_sample_dict(data, self.index.fields)
+
+    def _new_growing_epoch(self) -> int:
+        """Start a new growing epoch, in which we own the sample sequence because it grows.
+
+        Returns:
+            int: The epoch key, a handle which is given back to the caller.
+        """
+        with self._lock:
+            key = self._next_epoch_key
+            self._next_epoch_key += 1
+            epoch = list(self._loaded_epoch)
+            self._key_to_epoch[key] = epoch
+        return key
+
+    def _next_id(self, key: int) -> Optional[int]:
+        """Get next sample of the growing epoch given by key, or None if done.
+
+        If we are currently out of samples but not finished downloading the
+        shards, blocks until it has new samples.
+
+        Args:
+            key (int): The epoch key, a handle for this sequence.
+
+        Returns:
+            int: ID of next sample.
+        """
+        while True:
+            with self._lock:
+                epoch = self._key_to_epoch[key]
+                if epoch:
+                    return epoch.pop()
+                elif self._is_loading_complete:
+                    del self._key_to_epoch[key]
+                    return None
+                else:
+                    pass
+            sleep(0.25)
+
+    def _iter_ids(self) -> Iterator[int]:
+        """Get an iterator over all our sample IDs.
+
+        Returns:
+            Iterator[int]: Each sample ID.
+        """
+        with self._lock:
+            have_full_epoch = self._is_loading_complete
+
+        if have_full_epoch:
+            epoch = list(self._loaded_epoch)
```

Comment:
Make _loaded_epoch as np.array. Converting to python list and then passing it to np.random.shuffle may be expensive.

Reply:
Oof, benchmarked this and it turns out the shuffling and array creation are expensive at >1e6 samples

Proposed solution (shard shuffle then interleave into global) vs baseline (extend then global shuffle):

```
from argparse import ArgumentParser
from matplotlib import pyplot as plt
import numpy as np
from time import time
from tqdm import tqdm


def parse_args():
    args = ArgumentParser()
    args.add_argument('--num_shards', type=int, default=1000)
    args.add_argument('--approx_shard_size', type=int, default=1000)
    args.add_argument('--out', type=str, default='plot.png')
    return args.parse_args()


def extend_and_shuffle(samples_per_shard):
    shard_ends = samples_per_shard.cumsum()
    shard_begins = shard_ends - samples_per_shard

    t0 = time()
    tt = [0]
    nn = [0]
    ids = []
    for shard in tqdm(np.random.permutation(len(samples_per_shard))):
        begin = shard_begins[shard]
        end = shard_ends[shard]
        new_ids = list(range(begin, end))
        ids += new_ids
        np.random.shuffle(ids)
        t = time() - t0
        tt.append(t)
        n = len(ids)
        nn.append(n)
    return tt, nn


def shuffle_and_interleave(samples_per_shard):
    shard_ends = samples_per_shard.cumsum()
    shard_begins = shard_ends - samples_per_shard

    t0 = time()
    tt = [0]
    nn = [0]
    ids = None
    for shard in tqdm(np.random.permutation(len(samples_per_shard))):
        begin = shard_begins[shard]
        end = shard_ends[shard]
        new_ids = begin + np.random.permutation(end - begin)
        if ids is None:
            ids = new_ids
            continue
        num_cycles = len(new_ids)
        cycle_len = len(ids) // num_cycles
        cutoff = num_cycles * cycle_len
        old_block = ids[:cutoff].reshape(num_cycles, cycle_len)
        tail = ids[cutoff:]
        new_block = np.expand_dims(new_ids, 1)
        block = np.concatenate([old_block, new_block], 1)
        ids = np.concatenate([tail, block.flatten()])
        t = time() -t0
        tt.append(t)
        n = len(ids)
        nn.append(n)
    return tt, nn


def main(args):
    samples_per_shard = (args.approx_shard_size * np.exp(np.random.normal(0, 0.25, args.num_shards))).astype(np.int64)
    print(samples_per_shard)
    print()
    print('Shards: %d' % len(samples_per_shard))
    print()
    x = samples_per_shard
    print('Samples per shard:')
    print(f'* Min {x.min()}')
    print(f'* Mean {x.mean():.3f}')
    print(f'* Std {x.std():.3f}')
    print(f'* Max {x.max()}')
    print()

    plt.title('Shard Loading Speed on Simulated ImageNet')
    plt.xlabel('Time (seconds)')
    plt.ylabel('Samples')

    print('Bench extend_and_shuffle...')
    tt, nn = extend_and_shuffle(samples_per_shard)
    plt.plot(tt, nn, c='blue', label='extend + shuffle')

    print('Bench shuffle_and_interleave...')
    tt, nn = shuffle_and_interleave(samples_per_shard)
    plt.plot(tt, nn, c='green', label='shuffle + interleave')

    plt.savefig(args.out)


if __name__ == '__main__':
    main(parse_args())
    ```

Reply:
![plot](https://user-images.githubusercontent.com/2980246/162794227-391f9987-ca13-4cc4-ba79-1f1d26360710.png)

Reply:
How important is 40 seconds for epoch 1? From my understanding the processing of 1M samples will in ~64 separate threads, on high-end CPUs, and the download speed from S3 is far slower, O(minutes). So if we get perfect shuffling in 40 seconds, I think that's totally fine.

Reply:
Worth noting this is worst-case, in which no sample IDs are being popped off and used at the same time, so the sample ID list grows maximally. In real world usage, that limits the badness

The extend and shuffle is O(n^2) though right? Not tractable for much larger than ImageNet

Reply:
Agreed on this being the worst case, and will be better in practice as ids are being removed as samples are being processed.

Also, I think `np.random.shuffle` is O(n), the blue line is definitely leveling off, I think the underlying algo may be something like Fisher-Yates.

Overall as an ML-user, I would prefer the blue line with perfect shuffling, rather than the green line with imperfect shuffling across shards. I think there is also a 2x2 quadrant analysis here: If our buffers are growing to 100s of millions of samples, then the model must be going really slow relative to the dataloader, so the shuffling time doesn't make a difference. And if the model is going really fast, then IDs are being popped off really fast, so the shuffle list does not grow large.

Reply:
Sounds good, will not change to interleave

To be clear, leveling off means infinite time

Reply:
Oh oops I had my axes mixed up... 😂 but I think we're still OK, let's try it without interleave and revisit if we find a workload that is negatively impacted by the shuffling time.